[oneMKL][DFT] Full revision of DFT parts of specs + addition of {F,B}WD_STRIDES configuration parameters (#506)
oneapi-src/oneAPI-spec: source/elements/oneMKL/source/domains/dft/descriptor.rst

  ``oneapi::mkl::dft::precision::DOUBLE``) are referred to as "single-precision
  descriptors" (resp. "double-precision descriptors").

:ref:`onemkl_dft_enum_domain` dom
  Specifies the forward domain of the transform. The possible specialization
  values are ``oneapi::mkl::dft::domain::COMPLEX`` and
  ``oneapi::mkl::dft::domain::REAL``. Objects of the ``descriptor`` class
  specialized with :ref:`onemkl_dft_enum_domain` template parameter ``dom`` as
  value ``oneapi::mkl::dft::precision::COMPLEX`` (resp.
  ``oneapi::mkl::dft::precision::REAL``) are referred to as "complex
  descriptors" (resp. "real descriptors").

.. _onemkl_dft_descriptor_member_table:

.. rubric:: Descriptor class member functions

.. list-table::
   :header-rows: 1
   :widths: 30 70

   * -     Routines
     -     Description
   * -     :ref:`constructors<onemkl_dft_descriptor_constructor>`
     -     Creates and default-initializes a ``descriptor`` object for a
           :math:`d`-dimensional DFT of user-defined length(s)
           :math:`\lbrace n_1, \ldots, n_d\rbrace`.
   * -     :ref:`assignment operators<onemkl_dft_descriptor_assignment_operator>`

>      -     Self-explanatory.

Author: Done in [fda05259ad2bf98b5115cdaa631c5fb4178131f5](https://github.com/oneapi-src/oneAPI-spec/pull/506/commits/fda05259ad2bf98b5115cdaa631c5fb4178131f5)

   .. rubric:: Input Parameters

   length
      Length :math:`n_1 > 0` of the data sequence(s) for one-dimensional
      transform(s).

   lengths
      Vector of :math:`d > 0` lengths :math:`\lbrace n_1, \ldots, n_d\rbrace`
      of the data sequence(s) for :math:`d`-dimensional transform(s). The values
      are to be provided in that order and such that
      :math:`n_j > 0,\ \forall j \in \lbrace 1, \ldots, d \rbrace`.

   other
      Another ``descriptor`` object of the same type to copy or move.

.. container:: section

   .. rubric:: Throws

   The ``descriptor::descriptor()`` constructors shall throw the following
   exception if the associated condition is detected. An implementation may
   throw additional implementation-specific exception(s) in case of error
   conditions not covered here:

   :ref:`oneapi::mkl::host_bad_alloc()<onemkl_exception_host_bad_alloc>`
      If any memory allocations on host have failed, for instance due to
      insufficient memory.

   :ref:`oneapi::mkl::unimplemented()<onemkl_exception_unimplemented>`

>       If the dimension :math:`d`, *i.e.*, the size of vector ``lentghs``, is

Reviewer: ```suggestion
      If the dimension :math:`d`, *i.e.*, the size of vector ``lengths``, is
```

   implementation-specific exception(s) in case of error conditions not covered
   here:

   :ref:`oneapi::mkl::host_bad_alloc()<onemkl_exception_host_bad_alloc>`
      If any memory allocations on host have failed, for instance due to
      insufficient memory.

**Descriptor class member table:** :ref:`onemkl_dft_descriptor_member_table`

.. _onemkl_dft_descriptor_set_value:

set_value
+++++++++

The ``set_value`` member function of the ``descriptor`` class sets a
configuration value corresponding to a (read-write) configuration parameter for
the DFT(s) that a ``descriptor`` object defines. This function is to be used as
many times as required for all the necessary configuration parameters to be set
prior to committing the ``descriptor`` object (by calling its member function
:ref:`onemkl_dft_descriptor_commit`).

This function requires and expects exactly **two** arguments: it sets the
configuration value (second argument) corresponding to the configuration
parameter (first argument) ``param`` of type ``oneapi::mkl::dft::config_param``.
The type of the configuration value (second argument) to be set depends on the

> value of ``param``: it can be ``oneapi::mkl::dft::domain``,
> ``oneapi::mkl::dft::precision``, ``oneapi::mkl::dft::config_value`` or a native
> type like ``std::int64_t`` or ``float`` (more details available
> :ref:`here<onemkl_dft_enum_config_param>`).

Reviewer: ```suggestion
value of ``param``: it can be ``oneapi::mkl::dft::config_value`` or a native
type like ``std::int64_t`` or ``float`` (more details available
:ref:`here<onemkl_dft_enum_config_param>`).
```
I see where domain and precision came from in this list because `FORWARD_DOMAIN` and `PRECISION` are in that `config_param` table. But I think it's confusing to list them here since just the config_value enum or a native type are valid inputs to `set_value`.

Author: not to `get_value` though...

Author: My bad, I thought this was referring to something else, will change that.

Author: Addressed in [66791a78afc1ded03c9083a14391e29b36f9b056](https://github.com/oneapi-src/oneAPI-spec/pull/506/commits/66791a78afc1ded03c9083a14391e29b36f9b056)

.. _onemkl_dft_descriptor_get_value:

get_value
+++++++++

The ``get_value`` member function of the ``descriptor`` class queries the
configuration value corresponding to any configuration parameter for the DFT
that a ``descriptor`` object defines.

This function requires and expects exactly **two** arguments: it returns the
configuration value (into the element pointed by the second argument)
corresponding to the queried configuration parameter (first argument) ``param``
of type ``oneapi::mkl::dft::config_param``. The type of the second argument
depends on the value of ``param``: it is  a pointer to a writable element of
type ``oneapi::mkl::dft::domain``, ``oneapi::mkl::dft::precision``,
``oneapi::mkl::dft::config_value`` or a native type like ``std::int64_t`` or

> ``float`` (more details available :ref:`here<onemkl_dft_enum_config_param>`).

Author: Note added in [75af8a93c26153189dab3868c4875b6fbb0a0fdf](https://github.com/oneapi-src/oneAPI-spec/pull/506/commits/75af8a93c26153189dab3868c4875b6fbb0a0fdf). Please let me know if that's ok for you (thumbs-up) or comment.

   if the associated condition is detected. An implementation may throw
   additional implementation-specific exception(s) in case of error conditions
   not covered here:
   
   :ref:`oneapi::mkl::invalid_argument()<onemkl_exception_invalid_argument>`
      If the requested :ref:`onemkl_dft_enum_config_param` is not valid.

**Descriptor class member table:** :ref:`onemkl_dft_descriptor_member_table`

.. _onemkl_dft_descriptor_commit:

commit
++++++

The ``commit`` member function commits a ``descriptor`` object to the DFT
calculations it defines consistently with its configuration settings, by
completing all the initialization work (*e.g.*, algorithm selection, algorithm
tuning, choice of factorization, memory allocations, calculation of twiddle
factors, etc.) required by the chosen implementation for the desired DFT(s) on
the targeted device. Objects of the ``descriptor`` class **must** be committed
prior to using them in any call to :ref:`onemkl_dft_compute_forward` or
:ref:`onemkl_dft_compute_backward` (which trigger actual DFT calculations).

As specified :ref:`above<onemkl_dft_descriptor_set_value>`, all required

> configuration parameters must be set before this function is called. Any change
> in configuration operated on a ``descriptor`` object via a call to its
> :ref:`onemkl_dft_descriptor_set_value` member function *after* it was committed
> may not be successfully reflected by that object unless its ``commit`` member
> function is called again.

Reviewer: ```suggestion
configuration parameters must be set before this function is called. Any change
in configuration operated on a ``descriptor`` object via a call to its
:ref:`onemkl_dft_descriptor_set_value` member function *after* it was committed
results in an undefined state not suitable for computation until this ``commit`` member
function is called again.
```
I think we should be a little more forceful about disallowing any sort of configuration adjustment without subsequent re-commit.

      template <oneapi::mkl::dft::precision prec, oneapi::mkl::dft::domain dom>
      void descriptor<prec,dom>::commit(sycl::queue& queue);
   }

.. container:: section

   .. rubric:: Input Parameters

   queue 
      Valid ``sycl::queue`` object to which the operations relevant to the
      desired DFT(s) are to be enqueued.

.. container:: section

   .. rubric:: Throws

   The ``descriptor::commit()`` routine shall throw the following exceptions if
   the associated condition is detected. An implementation may throw additional
   implementation-specific exception(s) in case of error conditions not covered
   here (if the ``descriptor`` object's configuration was found to be
   inconsistent, for instance):
   
   :ref:`oneapi::mkl::invalid_argument()<onemkl_exception_invalid_argument>`
      If the queue is found to be invalid in any way.

   :ref:`oneapi::mkl::host_bad_alloc()<onemkl_exception_host_bad_alloc>`
      If any host side only memory allocations fail, for instance due to lack of
      memory.

   :ref:`oneapi::mkl::device_bad_alloc()<onemkl_exception_device_bad_alloc>`

>       If any device or shared memory allocation fail.

Reviewer: ```suggestion
      If any device or shared memory allocations fail.
```
